Cleanup property name computation

tautschnig opened the pull request:
This PR combines multiple cleanup steps (apologies for mixing them, but they depend on each other). Review commit-by-commit.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Make get_property_id more widely usable

1) Rename it to get_assertion_id as get_property_id already exists with a
slightly different semantics.
2) Move it from goto_symex_target_equationt to goto_trace_stept.
3) Use it when computing the property id in various *goto_trace implementations.

tautschnig committed:
Use a proper function identifier for lookups

The function name in a source location refers to the name used in the source
code, which need not coinincide with the name that we use internally.

This fix, together with the preceding commit, makes the source_location field of
conversion_dependencies unnecessary.

peterschrammel reviewed (approved):
I was contemplating storing the property ID in the goto step, but this works as well. Can we clarify in a comment the difference between property ID and assertion ID? - this seems still vague.

peterschrammel reviewed (changes requested):
The property id vs assertion id naming needs to be fixed.

tautschnig commented:
> The property id vs assertion id naming needs to be fixed.

Any concrete proposals beyond adding a comment? My thinking was that an assertion id generalises a property id, but I'm also ok to use "property" id everywhere.

peterschrammel commented:
The inconsistency comes from the following:
1. There are assertions that are in the goto program whose property ID is already attached to the source location of the corresponding assert instruction.
2. There are assertions that are not associated with an assert instruction in the goto program.

To accommodate the latter a `comment` field has been added to `SSA_stept` in order to hold the property description (which for the assertions of type 1 is also stored in the source location and just copied over to SSA_step in goto_symex). However, the propery_id is not stored in the SSA_stept, but later generated when printing (!) the goto_trace.

peterschrammel commented:
What we need is a consistent interface in `goto_trace_stept` and `SSA_stept` to get the property ID (and also its description aka 'comment') associated with an assertion. I think that is what your solution offers, but I don't think that it needs a new name.

Besides, storing the property_id in SSA_stept and goto_trace_stept alongside the `comment` would also make it more consistent and avoid regenerating the property ID in various places.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 20b1c1d0683b4303b20d9d4edc13d1a212f610b2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97920470

smowton commented:
A couple of the callers are clearly expecting a non-nil assertion ID -- could I suggest a `try_get` and `get` pair, where the latter asserts we found an ID?

tautschnig committed:
Clarify get_property_id's contract

get_property_id must only be called for an assert step, and such a step can only
be generated from either an ASSERT instruction or a backwards goto.

tautschnig commented:
@peterschrammel I have tried to implement your suggestions - is that what you were looking for?

@smowton I have clarified the interface contract, in a separate commit.

peterschrammel commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -1026,7 +1026,7 @@ irep_idt symex_target_equationt::SSA_stept::get_property_id() const
   else
   {
-    // return empty
+    UNREACHABLE;
```
Just a comment, no action required: If we want to introduce unwinding assertions for recursions then there would be another case here.

tautschnig replied:
Err, we do. I'll fix the code and will add a test.

tautschnig replied:
(Done)

peterschrammel reviewed (approved):
Thanks. That looks clearer now.

tautschnig committed:
Set a property id for recursion unwinding assertions

Previously we displayed an empty id.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 42ea75481616df017bb285783e392e26cdf4fd89).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97927228
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 212c793855b6a83d8eea337163f506d5a5d25e01).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97937840
